distributed/cfexecutor.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import print_function, division, absolute_import
+
+import weakref
+
+import six
+
+from toolz import merge
+
+from tornado import gen
+
+from concurrent.futures import Future, Executor, TimeoutError
+
+from .metrics import time
+from .utils import sync
+
+
+@gen.coroutine
+def _cascade_future(future, cf_future):
+    """
+    Coroutine that waits on future, then transmits its outcome to cf_future.
+    """
+    result = yield future._result(raiseit=False)
+    status = future.status
+    if status == 'finished':
+        cf_future.set_result(result)
+    elif status == 'cancelled':
+        cf_future.cancel()
+        # Necessary for wait() and as_completed() to wake up
+        cf_future.set_running_or_notify_cancel()
+    else:
+        try:
+            six.reraise(*result)
+        except BaseException as exc:
+            cf_future.set_exception(exc)
+
+
+@gen.coroutine
+def _wait_on_futures(futures):
+    for fut in futures:
+        try:
+            yield fut
+        except Exception:
+            pass
+
+
+class ClientExecutor(Executor):
+    """
+    A concurrent.futures Executor that executes tasks on a distributed Client.
+    """
+
+    _allowed_kwargs = frozenset(['pure', 'workers', 'resources', 'allow_other_workers'])
+
+    def __init__(self, client, **kwargs):
+        assert set(kwargs) <= self._allowed_kwargs, "unsupported kwargs to ClientExecutor"
```

Comment:
Perhaps an explicit `raise ValueError(...)` instead so that this sticks around even when assertions are turned off.